File: mobile/library/swift/src/stats/TimerImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of timer that can be used to track a distribution of time durations
@objcMembers
final class TimerImpl: NSObject, Timer {
  private let series: String
  private weak var engine: EnvoyEngine?

  init(elements: [Element], engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    super.init()
  }

  /// Record a new duration value for the distribution.
  func completeWithDuration(durationMs: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordHistogramDuration(self.series, durationMs: numericCast(durationMs))
  }
}

File: mobile/library/kotlin/io/envoyproxy/envoymobile/stats/TimerImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Timer` for time measurements e.g. distribution of durations.
 */
internal class TimerImpl : Timer {
  var envoyEngine: WeakReference<EnvoyEngine>
  var series: String
  var tags: Tags

  internal constructor(engine: EnvoyEngine, elements: List<Element>, tags: Tags = TagsBuilder().build()) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.series = elements.joinToString(separator = ".") { it.value }
    this.tags = tags
  }

  override fun recordDuration(durationMs: Int) {
    envoyEngine.get()?.recordHistogramDuration(series, this.tags.allTags(), durationMs)
  }

  override fun recordDuration(tags: Tags, durationMs: Int) {
    envoyEngine.get()?.recordHistogramDuration(series, tags.allTags(), durationMs)
  }
}
